### PR TITLE
[perf] separate impls for refs to simplifiable types

### PR DESCRIPTION
**Pull request by lqd**
As discussed on zulip, some of the coherence checks on the `window` crates are hot on `RefSimplifiedType`s, I believe because the crate contains a lot of types with `From` conversions between them (emulating `COM` hierarchies IIUC), including references to these types.

This PR separates impls for references to simplifiable types from the current non-blanket impls, for faster lookups.

Opening as draft for feedback purposes, especially as to what kind of tests @lcnr would like to see in this PR.

### CODE REVIEW

**Comment by lqd** [comment]
I'm not sure this is complete just yet or if it's still missing some important changes, so I'll leave it as draft for now, and I'd also like to add some tests.

But since this topic showed up in the discussion to reduce hotspots for the atypical `windows-rs` project, I've tried a non-incremental `cargo check --all-features` for the [`windows` crate](https://github.com/microsoft/windows-rs/tree/master/crates/libs/windows): the try build `8185d21406dfb1106547fc5cb973f6dffdc2b574` was around 9% faster (330s vs 360s) compared to its parent `731e0bf721c1ec2c7843547e86b6157b40a437d1`.

As shown in the perf run above, these patterns are quite uncommon in our benchmark suite, and I believe in the wider ecosystem, but they do stress coherence, and could be present in other similar big crates bindings to existing type hierarchies (e.g. to other OS APIs, maybe UI toolkits with OOP flavors, etc).

With that in mind, r? @lcnr for when they come back from vacation.

File: compiler/rustc_middle/src/ty/trait_def.rs [hunk 1]
```diff
@@ -147,8 +143,24 @@ impl<'tcx> TyCtxt<'tcx> {
...
+        // Non-blanket impls are grouped into two different maps:
+        // 1) impls for references to simplifiable types
+        if let TyKind::Ref(_, ref_ty, _) = self_ty.kind() {
+            if let Some(simplified_ref_ty) =
+                fast_reject::simplify_type(self, *ref_ty, TreatParams::AsInfer)
+            {
+                if let Some(impls) = impls.impls_for_ref_x.get(&simplified_ref_ty) {
+                    return impls.iter().copied();
+                }
+            }
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/trait_def.rs:156]: missing the `impls_for_ref_x` impls if you have `&T`. Doesn't matter with how this method is used but we should still be careful for it to be correct.

File: compiler/rustc_middle/src/ty/trait_def.rs [hunk 2]
```diff
@@ -177,25 +189,65 @@ impl<'tcx> TyCtxt<'tcx> {
...
+                        }
+                    }
+                }
+            } else {
+                for &impl_def_id in impls.non_blanket_impls.values().flatten() {
+                    if let result @ Some(_) = f(impl_def_id) {
+                        return result;
+                    }
                 }
             }
```
Comment (lcnr) [inline compiler/rustc_middle/src/ty/trait_def.rs:252]: the else branch is shared with `RefSimplifiedType`, isn't it? I think it makes sense to do
```rust 
if let Some(simp) = fast_reject::simplify_type(self, self_ty, TreatParams::AsPlaceholder) {
    impls.blanket_impls.iter()
     
    impls.non_blanket_impls.get(simp).iter()

    if let ty::Ref(_, ref_ty, mutablity) {
        if let Some(ref_simp) = fast_reject::simplify_type(self, *ref_ty, TreatParams::AsPlaceholder) {
            impls.impls_for_ref_x.get(ref_simp).iter()
        } else {
            impls.impls_for_ref_x.iter().flatten()
        }
    }
} else {
    self.all_impls(def_id)
}
```

**Review by lcnr — COMMENTED** [review]
a few nits, after that

r=me

**Comment by lcnr** [comment]
ah for tests:

probably:

`&_` correctly detecting impls for `&u32`.

coherence: `&T` overlapping with `&u32`

impls with `&u32` specializing an impl of `&T` and correctly using associated item definitions of the `&T` impl in the `&u32` impl.

That should cover all the edge cases I can think of rn

File: compiler/rustc_middle/src/ty/trait_def.rs (hunk 2 above)
Comment (lqd) [reply compiler/rustc_middle/src/ty/trait_def.rs:206]: Good idea. I've followed this suggestion when rebasing.

**Comment by lqd** [comment]
(rebased to discuss with lcnr, but the PR is still waiting on me)

@bors try

**Comment by lqd** [comment]
Sanity check on `windows-rs` with the try build `cd2704ff205c1cca87105183ec3fc80182b8f5bb` and its parent `871b5952023139738f72eba235063575062bc2e9`, on `full` and `incr-full` scenarios:

```
Benchmark 1: CARGO_INCREMENTAL=0 cargo +winrs-try-parent check --all-features
  Time (mean ± σ):     336.616 s ±  0.582 s    [User: 327.807 s, System: 12.608 s]
  Range (min … max):   335.954 s … 337.044 s    3 runs

Benchmark 2: CARGO_INCREMENTAL=0 cargo +winrs-try check --all-features
  Time (mean ± σ):     309.976 s ±  1.063 s    [User: 301.125 s, System: 12.671 s]
  Range (min … max):   309.017 s … 311.119 s    3 runs

Summary
  'CARGO_INCREMENTAL=0 cargo +winrs-try check --all-features' ran
    1.09 ± 0.00 times faster than 'CARGO_INCREMENTAL=0 cargo +winrs-try-parent check --all-features'
```

```
Benchmark 1: cargo +winrs-try-parent check --all-features
  Time (mean ± σ):     533.431 s ±  0.708 s    [User: 511.599 s, System: 25.910 s]
  Range (min … max):   532.816 s … 534.205 s    3 runs

Benchmark 2: cargo +winrs-try check --all-features
  Time (mean ± σ):     477.190 s ±  0.913 s    [User: 456.594 s, System: 24.689 s]
  Range (min … max):   476.411 s … 478.195 s    3 runs

Summary
  'cargo +winrs-try check --all-features' ran
    1.12 ± 0.00 times faster than 'cargo +winrs-try-parent check --all-features'
```

**Comment by lqd** [comment]
Coherence checks are on their way of being switched from the current solver to the next trait solver, via `-Ztrait-solver=next-coherence`. It performs the same in the rustc-perf benchmarks, and on `windows-rs`. 

However, this PR while being very beneficial to e.g. `windows-rs`, is very specific to the old solver architecture, and cannot be applied to the new solver. After discussing with @lcnr, we decided to close this PR because it will become obsolete as soon as the new solver is used for coherence, and landing prior to that, only for a short period, wouldn't be particularly useful.

**Pull request closed without merging**